### PR TITLE
(PUP-3483) Fix systemd provider to scan for changed units

### PR DESCRIPTION
**Pull request by vincentl-ipggroup**
Description:
Originally most people used to create an module based Exec to get a _systemd deamon-reload_, every time that a Systemd unit has been changed.
This could have the same exact command defined in multiple modules to ensure to get it run when required.

Suggested Fix:
This Commit will add a new function to check if an unit is requiring a deamon-reload before to (re)start it.
If the unit is flagged with **NeedDaemonReload=yes**, then a _systemd deamon-reload_ will be run.
If multiple units have been updated, the first one flagged will trigger the _deamon-reload_ for all of them, and the others will then be flagged **NeedDaemonReload=no**
So the command will run only once in a puppet run.

CPU Cost:
- Checking for the NeedDaemonReload=yes flag is costing around 6ms.
Running a single systemctl daemon-reload cost aroung 100ms.
Which means that even if you are **(re)starting** more than 100 units in your modules, this will not cost around **600ms to check them all**, and run a single systemctl daemon-reload if required, without to define a single Exec systemd deamon-reload in your modules. (~ 700 ms)
- In the current state, if you are **(re)starting** 7 modified Systemd units, they may call 7 specifics _daemon-reload_ Exec from multiples modules or shared accross them, and it will cost you ~ 700 ms.

### CODE REVIEW

File: lib/puppet/provider/service/systemd.rb [hunk 1]
```diff
@@ -123,6 +123,15 @@ def debian_enabled?
     end
   end
 
+  def daemon_reload?
```
Comment (vincentl-ipggroup) [inline lib/puppet/provider/service/systemd.rb]: Define the daemon_reload? function to check if the unit is requiring to trigger a "systemctl daemon-reload"
This function is called only on start & restart unit options.

**Comment by melissa** [comment]
Hi @vincentl-ipggroup, thank you for your contribution!

I'll try to review this within the next week.

**Comment by melissa** [comment]
Hey @vincentl-ipggroup 

We have to have some additional discussion internally on this. From some basic review, I'm not convinced puppet should take on this responsibility, as a few folks mentioned in the ticket. I'm kicking this up the chain to get help answering that question.

**Comment by vincentl-ipggroup** [comment]
> Hey @vincentl-ipggroup
> 
> We have to have some additional discussion internally on this. From some basic review, I'm not convinced puppet should take on this responsibility, as a few folks mentioned in the ticket. I'm kicking this up the chain to get help answering that question.

Hi Melissa, 

I can understand that you have to discuss it internally.
But actually if you check my [last comment in the ticket](https://tickets.puppetlabs.com/browse/PUP-3483?focusedCommentId=601120&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-601120), Puppet is already taking responsibility of doing some daemon-reload, as part of the "enable" action. 
Including this feature in the puppet code will not change the way it's already working, but improve it to ensure that the changes are fully applied, without to get some additional Exec commands.
And the responsibility of doing changes in the Systemd unit files will always be on the Admin who does it.

Furthermore, the modification I'm proposing is only applying to the Systemd Unit files managed by puppet.
___
_If an admin is manually changing a Systemd Unit file NOT managed by puppet, the "daemon-reload" will not be triggered if puppet has to (re)start an other Unit file.
The **"systemctl show @resource[:name] --property=NeedDaemonReload"** will only return the status of the Unit file that Puppet is going to (re)start. If there is no change, nothing will happened and the admin will be able to run a "daemon-reload" to apply his own change at its own time._
____

I hope that will help you to understand the purpose of this PR, and give you more material in your internal discussion.

**Comment by melissa** [comment]
Okay, it looks like enough folks want to see this in that I think it's a good idea to move forward on this. First though, I'd like to see a few updates.

Primarily, could you add some additional tests? We need a case covering what happens when `daemon_reload?` returns `yes`.

Additionally, could you add some documentation to the new method? I see you have a comment in the PR, but it would be good to have something along those lines in the code itself.

We should also squash your commits down into a single commit.

I'm happy to provide guidance or assistance with any of the above. Just ping me.

Thank you also for your patience on this. I very much appreciate it!

**Comment by vincentl-ipggroup** [comment]
Hi Melissa,

Good to know that other people are interested with this.

Just need to clarify some of the points:
- Do you want me to provide logs from tests, or do you want me to create a new Unit Test to cover it ?
- Where do you want me to add some documentation in ?  In the systemd.rb or soemwhere else ?

**Comment by melissa** [comment]
Hey Vincent

No need for logs, just create a new unit test to cover the new behavior.

And just add the documentation in comment form above your new method in lib/puppet/provider/service/systemd.rb. You can see some examples on other methods in that file.

Thank you!

**Comment by vincentl-ipggroup** [comment]
Ok, Cool.

I will try to get a unit test running for this, it will be a good exercise to improve my coding ^^
and then, I will add the comment in the systemd.rb file and squash the commits.

**Comment by melissa** [comment]
Thank you!

I'm glad you figured out the `--force` flag. Be careful with that, it changes history. In this case, that's exactly what we wanted to do since we were squashing commits. Just make sure it isn't your default!

The travis failure is interesting.  I don't think it's related to your change. I ran it again, but it failed in the same way. I'll poke at it this morning to see if I can figure out what is up.

Thank you for adding tests and comments! This looks great to me. As soon as we figure out Travis, I'll try to get this merged in!

**Comment by melissa** [comment]
This travis failure is super odd. I can't replicate it locally....

**Comment by melissa** [comment]
okay, I replicated it. It's something to do with the gem/ruby/bundle version combination
```
[1] Melissa@bubba:/Users/Melissa/puppet:(fix/master/PUP-3483)$ gem --version
2.7.8
[0] Melissa@bubba:/Users/Melissa/puppet:(fix/master/PUP-3483)$ ruby --version
ruby 2.4.0p0 (2016-12-24 revision 57164) [x86_64-darwin17]
[0] Melissa@bubba:/Users/Melissa/puppet:(fix/master/PUP-3483)$ bundle --version
Bundler version 1.16.6
[0] Melissa@bubba:/Users/Melissa/puppet:(fix/master/PUP-3483)$ gem --version
2.7.8
```

I downgraded the version of gem I was using locally
```
[0] Melissa@bubba:/Users/Melissa/puppet:(fix/master/PUP-3483)$ ruby --version
ruby 2.4.0p0 (2016-12-24 revision 57164) [x86_64-darwin17]
[0] Melissa@bubba:/Users/Melissa/puppet:(fix/master/PUP-3483)$ bundle --version
gBundler version 1.17.1
[0] Melissa@bubba:/Users/Melissa/puppet:(fix/master/PUP-3483)$ gem --version
2.6.8
```
and tests passed.

This is officially not an issue with your code.

**Comment by melissa** [comment]
jenkins please test this with ubuntu1604a

**Comment by vincentl-ipggroup** [comment]
Hi @melissa,

Thanks for chasing this issue and approving the PR.
I'm exciting to get it included in the Puppet code ^^
Waiting for the PR to be merged.

Thanks for your patience and your help to sort this out.

Cheers

**Comment by melissa** [comment]
Okay, we have the change we need in the master branch now (https://github.com/puppetlabs/puppet/commit/fec0ce498c8b72f6c4106596bffce4cc53e44f0e if you're curious).

Could you rebase against the master branch so you get that commit?  If you need any guidance, let me know!

**Comment by vincentl-ipggroup** [comment]
Hi @melissa 

I've rebased my branch from the Master branch.
Unfortunately, one fo the travis test failed with the error 
````
This commit summary didn't match CONTRIBUTING.md guidelines:
		(PA-2183) Update acceptance tests to work with Beaker 4
````

The second weird stuff is that all commits from master have been added to my Pull Request ? 
Is it a normal behaviour ? I just followed the Github guides [Configuring a remote for a fork](https://help.github.com/articles/configuring-a-remote-for-a-fork/) & [Syncing a fork](https://help.github.com/articles/syncing-a-fork/), and then pull my master and rebase my branch from it.

**Comment by melissa** [comment]
jenkins please test this on windows2012r2-64a

**Comment by melissa** [comment]
All checks have passed!

**Comment by ekohl** [comment]
I just noticed this was merged but I'm wondering why it's only done on start and restart. Stopping and reloading can also be affected because those instructions can be specified in a unit file.

**Comment by vincentl-ipggroup** [comment]
> I just noticed this was merged but I'm wondering why it's only done on start and restart. Stopping and reloading can also be affected because those instructions can be specified in a unit file.

Hi @ekohl,

Thanks for your interest on this Pull Request.
Actually, that's based on my choices, which I can explain.

First of all, the "reload" option is not yet  available for Systemd in Puppet. If you are managing service with reload option, you certainly use an exec to reload it when changing config files, ...
You can follow the Jira ticket [PUP-1054](https://tickets.puppetlabs.com/browse/PUP-1054) opened for this feature.

Regarding the daemon-reload before the "stop" action, I've been thinking of it.
But based on my own experience, nobody is changing a unit file and stopping it through Puppet at the same time.
I can't image someone updating a unit file which is going to notify the service to be stopped through puppet.
Usually you are changing your unit file which will notify the Unit for a restart. And when your config is applied, you may want to change the status of your Unit (to stopped).
My aim was to save some "useless" calls to the function, but I may be wrong.
Moreover, please remember that some "daemon-reload" actions are also performed when changing the "enabled/disable" Unit option. Which in this case is not going cover the point, as the Unit is stopped before changing the setting of the "enabled/disable" Unit option in Puppet.
````
/bin/systemctl stop <Unit>
/bin/systemctl disable <Unit>
````

I hope it give you a better view of my choices.
Feel free to share your thoughts, if you have some cases which can required to get the daemon-reload before to stop an Unit.

Cheers,
Vincent

**Comment by ekohl** [comment]
@vincentl-ipggroup I think you're right and that was also my gut feeling but I'm glad you thought about it more than just a gut feeling. Thanks for implementing this useful improvement!

**Pull request merged**